lib/Sema/ConstantnessSemaDiagnostics.cpp

```diff
@@ -0,0 +1,371 @@
+/// static member of the type.
+static void diagnoseError(Expr *errorExpr, const ASTContext &astContext,
+                          FuncDecl *funcDecl) {
+  DiagnosticEngine &diags = astContext.Diags;
+  Type exprType = errorExpr->getType();
+  SourceLoc errorLoc = errorExpr->getLoc();
+
+  // Diagnose atomics ordering related error here.
+  if (hasSemanticsAttr(funcDecl,
+                       semantics::ATOMICS_REQUIRES_CONSTANT_ORDERINGS)) {
+    NominalTypeDecl *nominalDecl = exprType->getNominalOrBoundGenericNominal();
+    if (!nominalDecl) {
+      // This case should normally not happen. This is a safe guard against
+      // possible mismatch between the atomics library and the compiler.
+      diags.diagnose(errorLoc, diag::argument_must_be_constant);
+    }
+    diags.diagnose(errorLoc, diag::atomics_ordering_must_be_constant,
+                   nominalDecl->getName());
+    return;
+  }
+
+  // Diagnose os_log specific errors here.
+
+  // Diagnose primitive stdlib types.
+  if (exprType->isBool()) {
+    diags.diagnose(errorLoc, diag::oslog_arg_must_be_bool_literal);
+    return;
+  }
+  if (isStringType(exprType)) {
+    diags.diagnose(errorLoc, diag::oslog_arg_must_be_string_literal);
+    return;
+  }
+  if (isIntegerType(exprType)) {
+    diags.diagnose(errorLoc, diag::oslog_arg_must_be_integer_literal);
+    return;
+  }
+  if (isFloatType(exprType)) {
+    diags.diagnose(errorLoc, diag::oslog_arg_must_be_float_literal);
+    return;
+  }
+  if (exprType->is<MetatypeType>()) {
+    diags.diagnose(errorLoc, diag::oslog_arg_must_be_metatype_literal);
+    return;
+  }
+  if (exprType->is<AnyFunctionType>()) {
+    diags.diagnose(errorLoc, diag::oslog_arg_must_be_closure);
+    return;
+  }
+  if (EnumDecl *enumDecl = exprType->getEnumOrBoundGenericEnum()) {
+    diags.diagnose(errorLoc, diag::oslog_arg_must_be_enum_case,
+                   enumDecl->getName());
+    return;
+  }
+  NominalTypeDecl *nominalDecl = exprType->getNominalOrBoundGenericNominal();
+  if (!nominalDecl) {
+    // This case should normally not happen. This is a safe guard against
+    // possible mismatch between the os overlay and the compiler.
+    diags.diagnose(errorLoc, diag::argument_must_be_constant);
+    return;
+  }
+  // If this is OSLogMessage, it should be a string-interpolation literal.
+  Identifier declName = nominalDecl->getName();
+  if (declName == astContext.Id_OSLogMessage) {
+    diags.diagnose(errorLoc, diag::oslog_message_must_be_string_interpolation);
+    return;
+  }
+  diags.diagnose(errorLoc, diag::oslog_arg_must_be_type_member_access,
+                 declName);
+}
+
+/// Given a call \c callExpr, if some or all of its arguments are required to be
+/// constants, check that property on the arguments.
+static void diagnoseConstantArgumentRequirementOfCall(const CallExpr *callExpr,
+                                                      const ASTContext &ctx) {
```

Comment:
Have you considered integrating this into a constraint solver? There is a way to check flags like this already when argument is matched to a parameter for a given overload choice and record a fix per argument which would diagnose as a constness violation. In general I think we shouldn't be producing solutions which are "incorrect" this way.

Reply:
@xedin I am worried about putting this in the main type checking algorithm as this is not a  language feature yet. We are just using some semantics attribute to enforce this on specific APIs. This uses hard coded knowledge of those APIs which are also outside the stdlib. The semantics attribute and also the constraints may also change as the APIs evolves (until of course we evolve  a language feature that can subsume this ad hoc check). I think it will be a language feature eventually, and the annotation will be generalized to a type or parameter annotation, at which point it should certainly be in the constraint solver. But now, it is more fine grained to specific APIs. That is why I wonder if it would be best as a MiscDiagnostics.

Reply:
We already have multiple precedents of that so I wouldn't worry that much. I'd say we should just add a flag and put everything into the solver for a couple of reasons:

1. We'd have to do this anyway :)
2. It's much easier to work with calls in the solver without involving expressions;
3. We are planning to change AST representation of calls so it would be great if new uses of TupleExpr/ParenExpr could be avoided;
4. Avoids producing incorrect solutions, which means that solver behavior when it comes to overloading would be correct for free;
5. Fix and diagnostics could be worked on separately for failure detection.

Reply:
@xedin Okay. I am quite new to that code base. It would very helpful  if you can provide some pointers to how I can implement this in the constraint solver.  I did a quick glance, and found somethings.  

I see that there is `TypeChecker::getDeclTypeCheckingSemantics`. I can extend that enum to include these functions. I see that it is used in `finishApply` (particularly in `finishApplyOfDeclWithSpecialTypeCheckingSemantics`), and in `getTypeOfReferenceWithSpecialTypeCheckingSemantics`. It seems to me that most of the  logic can go inside the `finishApplyOfDeclWithSpecialTypeCheckingSemantics`, where it calls the `coerceCallArguments` and then checks for constantness of some arguments. The diagnostics can also be emitted there. Does it match with what you were thinking?

Reply:
I think the place you are looking for is `matchCallArguments` https://github.com/apple/swift/blob/master/lib/Sema/CSSimplify.cpp#L967. That's the place we arguments are matched to their corresponding parameter types. `matchCallArguments` also retrieves selected overload used to form this call, so all you need to do is to determine which parameters have to be constants and fail such comparisons unless `shouldAttemptFixes()` is set. There are also a couple of examples of use of fixes available in that method. I'll be happy to help you out with any other questions you might have.